### PR TITLE
Replays

### PR DESCRIPTION
**Pull request by monster860**
Are you an admin who hates staring at logs? Do you want to look back at a round you have already played? Do you want to learn from the example of those that actually greentexted that round? Well, this PR is for you!

This pr logs literally everything that happens to a file. In a test merge, the file was 50 MB for a 1-hour cult round. (but putting it in a zipfile makes it 10x smaller), but for those 50 megabytes, you get all this information:
- The movements of every single object that exists on the station, including things inside backpacks, crates, etc.
- The appearances and changes in appearances of every object
- Lighting information is not included.
- Anything that gets sent to a player's chat (except for admin stuff)
- Which client inhabits which mob
- Where the fuck the hypospray was when you emagged the CMO's locker and it wasn't there, and you murdered the CMO and it wasn't on him either
- Where your objective target was the whole round when you couldn't find him
- What commit the game is running so that the viewer actually loads the right icons

file size isn't even a good concern not to do this, if it starts filling the server just have it copy the files to some cloud storage or something.

Performance concerns?
- SSdemo.mark_dirty, and it's friends, do not even show up on the profiler
- the actual file-writing part happens in SSdemo, which means if it's too slow, the subsystem thingy will tell it to finish up later.

A few caveats:
- Only things that call SSdemo.mark_dirty will update on the demo file.
- If the server is lagging, things won't update because it writes a bit of changes down every tick on SSdemo.

Of course we'll need to test this on the real server to see how large the file would actually be.

Anyways, on to actually viewing these files! I wrote a web based viewer for this really fast and dirty. It's not well optimized, but it works well enough for now.

Here's a video of that in action: https://streamable.com/w9bui

If this PR gets merged, I'll make a repo on the yogstation13 org and put the code on there, and help with integrating it into the website.

:cl: monster860, also asd for the quickwrite dll
add: Everything that happens gets logged to a replay file
/:cl:

### CODE REVIEW

**Comment by monster860** [comment]
The CI is wrong about /image not having invisibility. It does, it just isn't very useful most of the time.

**Comment by ghost** [comment]
Now I can kickstart my youtube ss13 career

**Comment by MegaEmpirical** [comment]
Fucking yes.

**Comment by Altoids1** [comment]
I don't know which is spammier -- having ``confidental`` be true by default or false.

**Comment by ThatLing** [comment]
> the actual file-writing part happens in SSdemo, which means if it's too slow, the subsystem thingy will tell it to finish up later.

What if the round ends and it hasn't finished writing yet?

**Comment by ThatLing** [comment]
As for the size, you could probably reduce by a lot of you made a proper format instead of just writing text

**Comment by monster860** [comment]
It will keep writing until the server closes.

**Comment by ThatLing** [comment]
> 
> It will keep writing until the server closes.

Yeah that's what I mean, won't the demo be corrupt if it's a laggy round and it hasn't finished writing when the server restart?

**Comment by monster860** [comment]
Not unless BYOND cuts off the end of a line

File: code/__DEFINES/subsystems.dm [hunk 1]
```diff
@@ -124,7 +125,7 @@
 #define RUNLEVELS_DEFAULT (RUNLEVEL_SETUP | RUNLEVEL_GAME | RUNLEVEL_POSTGAME)
 
 
-
+// Truly disgusting, TG. Truly disgusting.
```
Comment (optimumtact) [inline code/__DEFINES/subsystems.dm:128]: this is frankly a beautiful macro, and you should shut up

Comment (d3athrow) [reply code/__DEFINES/subsystems.dm:128]: uh oh

**Comment by HomocidlePancakes** [comment]
Holy shit

**Comment by Hopekz** [comment]
Really nice. Really good PR glad you're with us

**Comment by Codeatmos** [comment]
Holy shit they made moody blues into a fucking pr

File: code/__DEFINES/subsystems.dm (hunk 1 above)
Comment (optimumtact) [reply code/__DEFINES/subsystems.dm:128]: it's art, you wouldn't UNDERSTAND

Comment (d3athrow) [reply code/__DEFINES/subsystems.dm:128]: >art

more like stinky

**Comment by ghost** [comment]
Discord_5elSyPF7cX

https://github.com/MCHSL/byond-quickwrite/pull/1

**Comment by 81Denton** [comment]
This is very ambitious, hats off to you.

**Comment by Lostlarry** [comment]
did you fix the issue with the monitor contents being sideways?
image

**Comment by Manly-Dwarf** [comment]
you are a true legend, monster

**Comment by boodaliboo** [comment]
i am

**Comment by Quartich** [comment]
> did you fix the issue with the monitor contents being sideways?

Probably just the way it is loading the icons, and at least it isn't too much of a big deal

**Comment by SysadminJeroen** [comment]
I actually think that saving the lighting would be good, and it doesn't have to happen on every tile, only the status of the light sources (location and luminosity) so the client can recreate what the people could've seen.

Sometimes it is useful to know what people couldn't see.

**Comment by Ratvar** [comment]
Following AI doesn't follow it's eye, instead centers on core

**Comment by N3X15** [comment]
Is there a link to the quickwrite DLL source code?  We have some servers that run on Linux.

**Comment by optimumtact** [comment]
rust-g compiles for linux

**Comment by ghost** [comment]
@N3X15 https://github.com/MCHSL/byond-quickwrite

@optimumtact yeah but you gotta make a new function or edit the existing logging functions to not add timestamps

**Comment by optimumtact** [comment]
oh true sorry I keep forgetting about that

**Comment by Cyberboss** [comment]
I came up originally with this idea and never expected to happen.

I tip my fedora to you.

**Comment by ghost** [comment]
YALPER was never a secret, cyber

**Comment by d3athrow** [comment]
ive wanted to do this for 6 years

File: code/__DEFINES/subsystems.dm [hunk 2]
```diff
@@ -149,4 +150,6 @@
 			}\
 		}\
 		A.flags_1 &= ~OVERLAY_QUEUED_1;\
+		if(isturf(A)){SSdemo.mark_turf(A);}\
+		if(isobj(A) || ismob(A)){SSdemo.mark_dirty(A);}\
```
Comment (Cyberboss) [inline code/__DEFINES/subsystems.dm:154]: How much performance have we lost because this isn't an `else if` 😆

Comment (optimumtact) [reply code/__DEFINES/subsystems.dm:154]: none you fucking spoon, stop wasting time on meaningless microops that get lost in the noise

Comment (Cyberboss) [reply code/__DEFINES/subsystems.dm:154]: It gets called a LOT tbf

**Comment by Tharcoonvagh** [comment]
Now this is epic

**Pull request merged**